src/pubsub/src/publisher/publisher.rs

```diff
@@ -174,28 +176,111 @@ impl Worker {
     }
 
     async fn run(mut self) {
-        while let Some(msg) = self.rx.recv().await {
-            let client = self.client.clone();
-            let topic = self.topic_name.clone();
-            // In the future, we may also want to keep track of JoinHandles in order to
-            // flush the results.
-            let _handle = tokio::spawn(async move {
-                // For now, we just send the message immediately.
-                // We will want to batch these requests.
-                let request = client
-                    .publish()
-                    .set_topic(topic)
-                    .set_messages(vec![msg.msg]);
-
-                // Handle the response by extracting the message ID on success.
-                let result = request
-                    .send()
-                    .await
-                    .map(|mut response| response.message_ids.pop().unwrap_or_default());
-
-                // The user may have dropped the handle, so it is ok if this fails.
-                let _ = msg.tx.send(result);
-            });
+        let mut batch = Batch::new();
+        let delay = self.batching_options.delay_threshold;
+        let message_limit = self.batching_options.message_count_threshold;
+
+        let timer = tokio::time::sleep(delay);
+        // Pin the timer to the stack.
+        tokio::pin!(timer);
+        loop {
+            tokio::select! {
+                // Handle timer events.
+                // This branch will only be checked when there is a non-empty batch,
+                // so this will not fire continuously.
+                _ = &mut timer, if !batch.is_empty() => {
+                    let batch_to_send = std::mem::take(&mut batch);
+                    tokio::spawn(batch_to_send.send(self.client.clone(), self.topic_name.clone()));
+                }
+                // Handle receiving a message from the channel.
+                msg = self.rx.recv() => {
+                    match msg {
+                        Some(msg) => {
+                            // Reset the timer if this is the first message to be added to the batch.
+                            if batch.is_empty() {
+                                timer.as_mut().reset(tokio::time::Instant::now() + delay);
+                            }
+                            batch.push(msg);
+                            if batch.len() as u32 >= message_limit {
+                                let batch_to_send = std::mem::take(&mut batch);
+                                // In the future, we may also want to keep track of JoinHandles in order to
+                                // flush the results.
+                                let _handle =
+                                    tokio::spawn(batch_to_send.send(self.client.clone(), self.topic_name.clone()));
+                            }
```

Comment:
It feels like all the `tokio::spawn()` calls could be done inside a `batch.flush()` call?